TestPeerConnection_RoleSwitch fails

https://travis-ci.org/pion/webrtc/jobs/660184903
It fails again on rebuild.
```
=== RUN   TestPeerConnection_RoleSwitch
goroutine profile: total 30
4 @ 0x80758a4 0x804ca12 0x804c9ed 0x804c75c 0x82bb142 0x83006fc 0x82c07f9 0x809e321
#	0x82bb141	github.com/pion/transport/packetio.(*Buffer).Read+0x41		/go/pkg/mod/github.com/pion/transport@v0.8.10/packetio/buffer.go:135
#	0x83006fb	github.com/pion/webrtc/v2/internal/mux.(*Endpoint).Read+0x3b	/go/src/github.com/pion/webrtc/internal/mux/endpoint.go:36
#	0x82c07f8	github.com/pion/srtp.(*session).start.func1+0x98		/go/pkg/mod/github.com/pion/srtp@v1.2.7/session.go:124

4 @ 0x80758a4 0x8085981 0x808595c 0x80a5d66 0x8214719 0x8218430 0x83f0d48 0x809e321
#	0x808595b	sync.runtime_notifyListWait+0xfb				/usr/local/go/src/runtime/sema.go:510
#	0x80a5d65	sync.(*Cond).Wait+0x65						/usr/local/go/src/sync/cond.go:56
#	0x8214718	github.com/pion/sctp.(*Stream).ReadSCTP+0x88			/go/pkg/mod/github.com/pion/sctp@v1.7.6/stream.go:107
#	0x821842f	github.com/pion/datachannel.(*DataChannel).ReadDataChannel+0x3f	/go/pkg/mod/github.com/pion/datachannel@v1.4.16/datachannel.go:188
#	0x83f0d47	github.com/pion/webrtc/v2.(*DataChannel).readLoop+0x87		/go/src/github.com/pion/webrtc/datachannel.go:319

2 @ 0x80758a4 0x804ca12 0x804c9ed 0x804c75c 0x82bb142 0x82f5705 0x8300e6f 0x809e321
#	0x82bb141	github.com/pion/transport/packetio.(*Buffer).Read+0x41		/go/pkg/mod/github.com/pion/transport@v0.8.10/packetio/buffer.go:135
#	0x82f5704	github.com/pion/ice.(*Conn).Read+0x74				/go/pkg/mod/github.com/pion/ice@v0.7.10/transport.go:78
#	0x8300e6e	github.com/pion/webrtc/v2/internal/mux.(*Mux).readLoop+0x9e	/go/src/github.com/pion/webrtc/internal/mux/mux.go:106

2 @ 0x80758a4 0x804ca12 0x804c9ed 0x804c75c 0x82bb142 0x83006fc 0x829aed9 0x809e321
#	0x82bb141	github.com/pion/transport/packetio.(*Buffer).Read+0x41		/go/pkg/mod/github.com/pion/transport@v0.8.10/packetio/buffer.go:135
#	0x83006fb	github.com/pion/webrtc/v2/internal/mux.(*Endpoint).Read+0x3b	/go/src/github.com/pion/webrtc/internal/mux/endpoint.go:36
#	0x829aed8	github.com/pion/dtls/v2.(*Conn).inboundLoop+0x98		/go/pkg/mod/github.com/pion/dtls/v2@v2.0.0-rc.7/conn.go:583

2 @ 0x80758a4 0x804ca12 0x804c9ed 0x804c77c 0x821800b 0x8217fea 0x84045e5 0x809e321
#	0x821800a	github.com/pion/sctp.(*Association).AcceptStream+0x3a			/go/pkg/mod/github.com/pion/sctp@v1.7.6/association.go:1123
#	0x8217fe9	github.com/pion/datachannel.Accept+0x19					/go/pkg/mod/github.com/pion/datachannel@v1.4.16/datachannel.go:129
#	0x84045e4	github.com/pion/webrtc/v2.(*SCTPTransport).acceptDataChannels+0x64	/go/src/github.com/pion/webrtc/sctptransport.go:143

2 @ 0x80758a4 0x804ca12 0x804c9ed 0x804c77c 0x82bd3c0 0x8432dd7 0x809e321
#	0x82bd3bf	github.com/pion/srtp.(*SessionSRTCP).AcceptStream+0x3f			/go/pkg/mod/github.com/pion/srtp@v1.2.7/session_srtcp.go:76
#	0x8432dd6	github.com/pion/webrtc/v2.(*PeerConnection).drainSRTP.func3+0x86	/go/src/github.com/pion/webrtc/peerconnection.go:980

2 @ 0x80758a4 0x804ca12 0x804c9ed 0x804c77c 0x82be030 0x8432b57 0x809e321
#	0x82be02f	github.com/pion/srtp.(*SessionSRTP).AcceptStream+0x3f			/go/pkg/mod/github.com/pion/srtp@v1.2.7/session_srtp.go:77
#	0x8432b56	github.com/pion/webrtc/v2.(*PeerConnection).drainSRTP.func2+0x46	/go/src/github.com/pion/webrtc/peerconnection.go:960

2 @ 0x80758a4 0x8070524 0x806f9fb 0x80be807 0x80bf396 0x80bf37b 0x81bb7ef 0x81cf63f 0x81ce407 0x82ede10 0x809e321
#	0x806f9fa	internal/poll.runtime_pollWait+0x4a			/usr/local/go/src/runtime/netpoll.go:184
#	0x80be806	internal/poll.(*pollDesc).wait+0x36			/usr/local/go/src/internal/poll/fd_poll_runtime.go:87
#	0x80bf395	internal/poll.(*pollDesc).waitRead+0x165		/usr/local/go/src/internal/poll/fd_poll_runtime.go:92
#	0x80bf37a	internal/poll.(*FD).ReadFrom+0x14a			/usr/local/go/src/internal/poll/fd_unix.go:219
#	0x81bb7ee	net.(*netFD).readFrom+0x3e				/usr/local/go/src/net/fd_unix.go:208
#	0x81cf63e	net.(*UDPConn).readFrom+0x3e				/usr/local/go/src/net/udpsock_posix.go:47
#	0x81ce406	net.(*UDPConn).ReadFrom+0x46				/usr/local/go/src/net/udpsock.go:121
#	0x82ede0f	github.com/pion/ice.(*candidateBase).recvLoop+0xef	/go/pkg/mod/github.com/pion/ice@v0.7.10/candidate_base.go:92

2 @ 0x80758a4 0x8083e59 0x81fb413 0x809e321
#	0x81fb412	github.com/pion/sctp.(*Association).writeLoop+0x122	/go/pkg/mod/github.com/pion/sctp@v1.7.6/association.go:456

2 @ 0x80758a4 0x8083e59 0x829901f 0x81fb13c 0x809e321
#	0x829901e	github.com/pion/dtls/v2.(*Conn).Read+0xae		/go/pkg/mod/github.com/pion/dtls/v2@v2.0.0-rc.7/conn.go:285
#	0x81fb13b	github.com/pion/sctp.(*Association).readLoop+0x13b	/go/pkg/mod/github.com/pion/sctp@v1.7.6/association.go:421

2 @ 0x80758a4 0x8083e59 0x82f854f 0x809e321
#	0x82f854e	github.com/pion/ice.(*Agent).startConnectivityChecks.func1.1+0xae	/go/pkg/mod/github.com/pion/ice@v0.7.10/agent.go:612

1 @ 0x80758a4 0x804ca12 0x804c9ed 0x804c75c 0x8110f5b 0x8113daa 0x8110c2a 0x8111fcd 0x811127b 0x844de44 0x80754d2 0x809e321
#	0x8110f5a	testing.(*T).Run+0x2ca		/usr/local/go/src/testing/testing.go:961
#	0x8113da9	testing.runTests.func1+0x59	/usr/local/go/src/testing/testing.go:1202
#	0x8110c29	testing.tRunner+0x99		/usr/local/go/src/testing/testing.go:909
#	0x8111fcc	testing.runTests+0x22c		/usr/local/go/src/testing/testing.go:1200
#	0x811127a	testing.(*M).Run+0x13a		/usr/local/go/src/testing/testing.go:1117
#	0x844de43	main.main+0x103			_testmain.go:322
#	0x80754d1	runtime.main+0x201		/usr/local/go/src/runtime/proc.go:203

1 @ 0x80758a4 0x804ca12 0x804c9ed 0x804c75c 0x82bb142 0x8402cc6 0x8402c9d 0x840a377 0x840a3ff 0x840a985 0x83fc0ba 0x809e321
#	0x82bb141	github.com/pion/transport/packetio.(*Buffer).Read+0x41		/go/pkg/mod/github.com/pion/transport@v0.8.10/packetio/buffer.go:135
#	0x8402cc5	github.com/pion/srtp.(*ReadStreamSRTP).Read+0x55		/go/pkg/mod/github.com/pion/srtp@v1.2.7/stream_srtp.go:69
#	0x8402c9c	github.com/pion/webrtc/v2.(*RTPReceiver).readRTP+0x2c		/go/src/github.com/pion/webrtc/rtpreceiver.go:155
#	0x840a376	github.com/pion/webrtc/v2.(*Track).Read+0xb6			/go/src/github.com/pion/webrtc/track.go:97
#	0x840a3fe	github.com/pion/webrtc/v2.(*Track).ReadRTP+0x5e			/go/src/github.com/pion/webrtc/track.go:103
#	0x840a984	github.com/pion/webrtc/v2.(*Track).determinePayloadType+0x34	/go/src/github.com/pion/webrtc/track.go:198
#	0x83fc0b9	github.com/pion/webrtc/v2.(*PeerConnection).startReceiver+0xe9	/go/src/github.com/pion/webrtc/peerconnection.go:784

1 @ 0x80758a4 0x8083e59 0x8420f6a 0x84221da 0x8110c2a 0x809e321
#	0x8420f69	github.com/pion/webrtc/v2.sendVideoUntilDone+0x129		/go/src/github.com/pion/webrtc/peerconnection_renegotation_test.go:19
#	0x84221d9	github.com/pion/webrtc/v2.TestPeerConnection_RoleSwitch+0x4d9	/go/src/github.com/pion/webrtc/peerconnection_renegotation_test.go:195
#	0x8110c29	testing.tRunner+0x99						/usr/local/go/src/testing/testing.go:909

1 @ 0x814626c 0x81460da 0x8143133 0x83ee9d7 0x809e321
#	0x814626b	runtime/pprof.writeRuntimeProfile+0x7b			/usr/local/go/src/runtime/pprof/pprof.go:708
#	0x81460d9	runtime/pprof.writeGoroutine+0x79			/usr/local/go/src/runtime/pprof/pprof.go:670
#	0x8143132	runtime/pprof.(*Profile).WriteTo+0x2c2			/usr/local/go/src/runtime/pprof/pprof.go:329
#	0x83ee9d6	github.com/pion/transport/test.TimeOut.func1+0x56	/go/pkg/mod/github.com/pion/transport@v0.8.10/test/util.go:18

panic: timeout

goroutine 1695 [running]:
github.com/pion/transport/test.TimeOut.func1()
	/go/pkg/mod/github.com/pion/transport@v0.8.10/test/util.go:21 +0xda
created by time.goFunc
	/usr/local/go/src/time/sleep.go:168 +0x37
FAIL	github.com/pion/webrtc/v2	87.472s
```

--- a/peerconnection.go
+++ b/peerconnection.go
@@ -29,6 +29,9 @@ type PeerConnection struct {
 	statsID string
 	mu      sync.RWMutex
 
+	// Negotiations must be processed serially
+	negotationLock sync.Mutex
+
 	configuration Configuration
 
 	currentLocalDescription  *SessionDescription
@@ -1536,6 +1539,9 @@ func (pc *PeerConnection) GetStats() StatsReport {
 
 // Start all transports. PeerConnection now has enough state
 func (pc *PeerConnection) startTransports(iceRole ICERole, dtlsRole DTLSRole, remoteUfrag, remotePwd, fingerprint, fingerprintHash string, currentTransceivers []*RTPTransceiver, incomingTracks map[uint32]trackDetails) {
+	pc.negotationLock.Lock()
+	defer pc.negotationLock.Unlock()
+
 	// Start the ice transport
 	err := pc.iceTransport.Start(
 		pc.iceGatherer,
@@ -1569,6 +1575,9 @@ func (pc *PeerConnection) startTransports(iceRole ICERole, dtlsRole DTLSRole, re
 }
 
 func (pc *PeerConnection) startRenegotation(currentTransceivers []*RTPTransceiver) {
+	pc.negotationLock.Lock()
+	defer pc.negotationLock.Unlock()
+
 	trackDetails := trackDetailsFromSDP(pc.log, pc.RemoteDescription().parsed)
 	for _, t := range currentTransceivers {
 		if t.Receiver() == nil || t.Receiver().Track() == nil {

--- a/peerconnection_renegotation_test.go
+++ b/peerconnection_renegotation_test.go
@@ -170,15 +170,7 @@ func TestPeerConnection_RoleSwitch(t *testing.T) {
 		onTrackFiredFunc()
 	})
 
-	connected, connectedFunc := context.WithCancel(context.Background())
-	pcFirstOfferer.OnConnectionStateChange(func(c PeerConnectionState) {
-		if c == PeerConnectionStateConnected {
-			connectedFunc()
-		}
-	})
-
 	assert.NoError(t, signalPair(pcFirstOfferer, pcSecondOfferer))
-	<-connected.Done()
 
 	// Add a new Track to the second offerer
 	// This asserts that it will match the ordering of the last RemoteDescription, but then also add new Transceivers to the end

